### PR TITLE
feat: Test with both go-ipfs and js-ipfs

### PR DESCRIPTION
**Pull request by ukstv**
Make CircleCI test using go-ipfs and js-ipfs based on `IPFS_FLAVOR` env variable. Uses go-ipfs by default, but if `IPFS_FLAVOR` env is set to `js` or `JS`, then `js-ipfs` is used.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -15,8 +15,17 @@ jobs:
...
           command: npm run test
           no_output_timeout: 30m
+          environment:
+            IPFS_FLAVOR: go
+      - run:
+          name: Test with js-ipfs
+          command: npm run test
+          no_output_timeout: 30m
+          environment:
+            IPFS_FLAVOR: js
```
Comment (oed) [inline .circleci/config.yml:28]: Can we make these two tasks run in parallel?

File: packages/canary-integration/src/create-ipfs.ts [hunk 2]
```diff
@@ -2,21 +2,33 @@ import * as dagJose from 'dag-jose'
 import { path } from 'go-ipfs'
 import * as Ctl from 'ipfsd-ctl'
 import * as ipfsHttp from 'ipfs-http-client'
-import type { IPFS, Options } from 'ipfs-core'
+import type { Options } from 'ipfs-core'
+import { create } from 'ipfs-core'
 import getPort from 'get-port'
 import mergeOpts from 'merge-options'
+import type { IpfsApi } from '@ceramicnetwork/common'
```
Comment (oed) [inline packages/canary-integration/src/create-ipfs.ts:9]: Is this not possible to import directly from js-ipfs yet?

**Review by oed — COMMENTED** [review]
Would be good if we can run both of these in parallel!

File: .circleci/config.yml (hunk 1 above)
Comment (oed) [reply .circleci/config.yml:28]: I guess we could create two different "jobs"?

**Review by stbrody — COMMENTED** [review]
my only comment is the same as joels - can we make the jobs parallel?  otherwise I think it looks great, though I'm still hoping we can de-dupe the two different paths for launching ipfs to reduce code duplication, but that can happen in a separate pr

**Comment by ukstv** [comment]
> I'm still hoping we can de-dupe the two different paths for launching ipfs to reduce code duplication

Sorry, could you elaborate please? I remember no previous discussion about de-duping anything.

File: packages/canary-integration/src/create-ipfs.ts (hunk 2 above)
Comment (ukstv) [reply packages/canary-integration/src/create-ipfs.ts:9]: It is possible, but for the sake of uniformity, I retained `IpfsApi`, until it is gone everywhere at once.

**Comment by stbrody** [comment]
> > I'm still hoping we can de-dupe the two different paths for launching ipfs to reduce code duplication
> 
> @stbrody Sorry, could you elaborate please? I remember no previous discussion about de-duping anything.

it was a comment on the previous PR.  We have code to create an ipfs node in both the `core` and `canary-integration` pacakges.  Would be good to reuse one version and get rid of the code duplication

File: .circleci/config.yml (hunk 1 above)
Comment (ukstv) [reply .circleci/config.yml]: it works https://app.circleci.com/pipelines/github/ceramicnetwork/js-ceramic/4772/workflows/a56b2290-5fe5-4386-88bb-abb74552a6f4

File: packages/canary-integration/src/create-ipfs.ts (hunk 2 above)
Comment (ukstv) [reply packages/canary-integration/src/create-ipfs.ts:9]: Relevant: https://github.com/ceramicnetwork/js-ceramic/issues/1937

**Comment by ukstv** [comment]
@stbrody tracked in https://github.com/ceramicnetwork/js-ceramic/issues/1938

Out of scope here, IMO. I know it is not DRY yet.

File: .circleci/config.yml (hunk 1 above)
Comment (oed) [reply .circleci/config.yml]: Nice!

Comment (oed) [reply .circleci/config.yml]: I wonder if there's a way to not duplicate the clone and build. With doing them in parallel we spend more CPU time but less wall clock time.

Comment (ukstv) [reply .circleci/config.yml]: In theory, you could do a single `build` job, and then run two ('with-go-ipfs', 'with-js-ipfs') jobs in parallel. In practice (prior experience+the internets on CircleCI specifics), with node packages, it gets slow in every dimension. Every CircleCI job executes in own container (or machine, or whatever executor). In order to pass data, like our fully installed and built repo, between jobs, you have to archive that, pass to storage, then download it and dearchive. For thousands of small files in node_modules, the process is far from perfect.

Comment (oed) [reply .circleci/config.yml]: ok, makes sense. Then let's keep as is :)

**Review by oed — APPROVED** [review]

File: .circleci/config.yml [hunk 3]
```diff
@@ -1,24 +1,41 @@
...
       - run:
+          name: Test
           command: npm run test
           no_output_timeout: 30m
+jobs:
+  with-go-ipfs:
+    machine:
+      image: ubuntu-2004:202107-02
+    environment:
+      NODE_OPTIONS: '--max_old_space_size=4096'
```
Comment (stbrody) [inline .circleci/config.yml:23]: nit: can probably remove this option from go-ipfs

**Review by stbrody — APPROVED** [review]

File: .circleci/config.yml (hunk 3 above)
Comment (ukstv) [reply .circleci/config.yml]: ok, will see, if it works :)

Comment (ukstv) [reply .circleci/config.yml]: it does.

Comment (stbrody) [reply .circleci/config.yml]: yeah, `NODE_OPTIONS` is an env var used by the nodejs runtime, it's not ipfs-specific at all, so I can't imagine it having any affect at all on a go-ipfs process

**Comment by stbrody** [comment]
I just realized that this PR doesn't update https://github.com/ceramicnetwork/js-ceramic/blob/e037f8ea8b6ae0fdd83a8e5d8d435ba89554b7ed/packages/cli/src/__tests__/create-ipfs.ts#L10, which means the ceramic-daemon tests aren't using go-ipfs.  Those are fairly large and extensive tests, so we'll definitely need to get them switched over to go-ipfs as well.  I'll file a ticket

**Pull request merged**